src/realm/object-store/sync/app_user.cpp

```diff
@@ -0,0 +1,311 @@
+#include <realm/object-store/sync/impl/app_metadata.hpp>
+#include <realm/object-store/sync/impl/sync_file.hpp>
+#include <realm/object-store/sync/mongo_client.hpp>
+#include <realm/object-store/sync/sync_manager.hpp>
+
+namespace realm::app {
+
+UserIdentity::UserIdentity(const std::string& id, const std::string& provider_type)
+    : id(id)
+    , provider_type(provider_type)
+{
+}
+
+User::User(Private, std::shared_ptr<app::App> app, std::string_view user_id)
+    : m_app(std::move(app))
+    , m_app_id(m_app->config().app_id)
+    , m_user_id(user_id)
+{
+    m_app->register_sync_user(*this);
+}
+
+User::~User()
+{
+    if (m_app) {
+        m_app->unregister_sync_user(*this);
+    }
+}
+
+std::string User::user_id() const noexcept
+{
+    return m_user_id;
+}
+
+std::string User::app_id() const noexcept
+{
+    return m_app_id;
+}
+
+std::vector<std::string> User::legacy_identities() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return m_data.legacy_identities;
+}
+
+std::string User::access_token() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return m_data.access_token.token;
+}
+
+std::string User::refresh_token() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return m_data.refresh_token.token;
+}
+
+SyncUser::State User::state() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    if (!m_app)
+        return SyncUser::State::Removed;
+    return m_data.access_token ? SyncUser::State::LoggedIn : SyncUser::State::LoggedOut;
+}
+
+bool User::is_anonymous() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return do_is_anonymous();
+}
+
+bool User::do_is_anonymous() const
+{
+    return m_data.access_token && m_data.identities.size() == 1 &&
+           m_data.identities[0].provider_type == app::IdentityProviderAnonymous;
+}
+
+std::string User::device_id() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return m_data.device_id;
+}
+
+bool User::has_device_id() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return !m_data.device_id.empty() && m_data.device_id != "000000000000000000000000";
+}
+
+UserProfile User::user_profile() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return m_data.profile;
+}
+
+std::vector<UserIdentity> User::identities() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return m_data.identities;
+}
+
+std::optional<bson::BsonDocument> User::custom_data() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return m_data.access_token.user_data;
+}
+
+std::shared_ptr<App> User::app() const
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return m_app;
+}
+
+SyncManager* User::sync_manager()
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return m_app ? m_app->sync_manager().get() : nullptr;
+}
+
+app::MongoClient User::mongo_client(const std::string& service_name)
+{
+    util::CheckedLockGuard lock(m_mutex);
+    return app::MongoClient(shared_from_this(), m_app->app_service_client(), service_name);
```

Comment:
should we cache the client?

Reply:
This constructor and the copy constructor do the exact same thing, so it wouldn't save any work.